File: test/router_test.php
<?php
require_once 'PHPUnit/Framework.php';

// had to install a bunch of stuff
// PEAR:
// http://www.phpunit.de/manual/current/en/installation.html
// http://www.newmediacampaigns.com/page/install-pear-phpunit-xdebug-on-macosx-snow-leopard
// and making sure that the pear gets installed in /usr/local
// also editig the /etc/php.ini file was tricky, the correct include_path is
// sh-3.2# grep 'include_path' /etc/php.ini | grep -v ';'
// include_path = ".:/usr/local/PEAR:/usr/lib/php/pear:/php/includes"
// then apachectl restart
 
// to run test
// $ phpunit --verbose foo.php 

/// http://www.phpunit.de/manual/current/en/api.html#api.assert

include(dirname(__FILE__).'/../lib/router.php');

class RouterTest extends PHPUnit_Framework_TestCase
{
	// default layout
	public function test_default_layout_yellow_page_because_about_layout_doesnt_exist()
	{
	  $router = new Router('/about/me','/index.php',dirname(__FILE__).'/test_app_default');
	  $router->route();
	  $this->assertEquals('default', $router->layout);
	  $this->assertEquals('me', $router->page);
	}
	public function test_default_layout_index_page()
	{
	 $router = new Router('/','/index.php',dirname(__FILE__).'/test_app_default');
	 $router->route();
	 $this->assertEquals('default', $router->layout);
	 $this->assertEquals('index', $router->page);
	}
	public function test_default_layout_default_page()
	{
	 $router = new Router('/default','/index.php',dirname(__FILE__).'/test_app_default');
	 $router->route();
	 $this->assertEquals('default', $router->layout);
	 $this->assertEquals('index', $router->page);
	}
	// about layout
	public function test_about_layout_me_page()
	{
	 $router = new Router('/about/me','/index.php',dirname(__FILE__).'/test_app_about');
	 $router->route();
	 $this->assertEquals('about', $router->layout);
	 $this->assertEquals('me', $router->page);
	}
  public function test_about_layout_index_page()
  {
   $router = new Router('/about','/index.php',dirname(__FILE__).'/test_app_about');
   $router->route();
   $this->assertEquals('about', $router->layout);
   $this->assertEquals('index', $router->page);
  }
	public function test_about_layout_yellow_page_animals()
	{
	 $router = new Router('/animals/dogs/brian','/index.php',dirname(__FILE__).'/test_app_default');
	 $router->route();
	 $this->assertEquals('default', $router->layout);
	 $this->assertEquals('dogs', $router->page);
	}
	// subdir default layout
	public function test_subdir_default_layout_contact_page()
	{
	 $router = new Router('/subdir/contact','/index.php',dirname(__FILE__).'/test_app_subdir_default');
	 $router->route();
	 $this->assertEquals('default', $router->layout);
	 $this->assertEquals('contact', $router->page);
	}
	public function test_subdir_about_layout_me_page_no_about_layout()
	{
	 $router = new Router('/subdir/about/me','/index.php',dirname(__FILE__).'/test_app_subdir_default');
	 $router->route();
	 $this->assertEquals('default', $router->layout);
	 $this->assertEquals('me', $router->page);
	}
	// subdir about layout
	public function test_subdir_about_layout_index_page()
	{
	 $router = new Router('/subdir/about','/index.php',dirname(__FILE__).'/test_app_subdir_default');
	 $router->route();
	 $this->assertEquals('default', $router->layout);
	 $this->assertEquals('about', $router->page);
	}
	public function test_subdir_about_layout_me_page_found_about_layout()
	{
	 $router = new Router('/subdir/about/me','/index.php',dirname(__FILE__).'/test_app_subdir_about');
	 $router->route();
	 $this->assertEquals('about', $router->layout);
	 $this->assertEquals('me', $router->page);
	}
	public function test_subdir_about_layout_index_page_found_about_layout()
	{
	 $router = new Router('/subdir/about','/index.php',dirname(__FILE__).'/test_app_subdir_about');
	 $router->route();
	 $this->assertEquals('about', $router->layout);
	 $this->assertEquals('index', $router->page);
	}
}

?>

File: test/router_test.rb
#!/usr/bin/env ruby

class Router
  attr_reader :layout, :page
  def initialize(request_uri, script_name, root)
    @request_uri = request_uri
    @script_name = script_name
    @root = root
  end
  
  def route!
    args = @request_uri.split('/')
    if args.length == 0    # '/'.split('/') => []
      @layout = 'default'
      @page = 'index'
    elsif args.length == 2 # '/about'.split('/') => ["", "about"]
      @layout = 'default'
      @page = args[1]
    elsif args # '/about/me'.split('/') => ["", "about", "me"]
      @layout = args[1]
      path = File.expand_path("#{@root}/layouts/#{@layout}.php")
      if File.exist?(path)
        # if about layout exists use it
      else
        @layout = 'default' # else fallback to the default layout
      end
      @page = args[2]
    end
  end
end

# http://php-template.local/gallery/dogs/yellow
# $_SERVER['REQUEST_URI']: /gallery/dogs/yellow
# $_SERVER['SCRIPT_NAME']: /index.php
# 
# http://php-template.local/
# $_SERVER['REQUEST_URI']: /
# $_SERVER['SCRIPT_NAME']: /index.php
#
# http://php-template.local/about/yellow
# $_SERVER['REQUEST_URI']: /about/yellow
# $_SERVER['SCRIPT_NAME']: /index.php
describe Router do
  it "default layout yellow page because about layout doesnt exist" do
    @router = Router.new("/about/yellow","/index.php", File.dirname(__FILE__)+"/test_app_default")
    @router.route!
    @router.layout.should == "default"
    @router.page.should == "yellow"
  end
  it "default layout index page" do
    @router = Router.new("/","/index.php", File.dirname(__FILE__)+"/test_app_default")
    @router.route!
    @router.layout.should == "default"
    @router.page.should == "index"
  end
  it "default layout default page" do
    @router = Router.new("/default","/index.php", File.dirname(__FILE__)+"/test_app_default")
    @router.route!
    @router.layout.should == "default"
    @router.page.should == "default"
  end
  it "default layout about page" do
    @router = Router.new("/about","/index.php", File.dirname(__FILE__)+"/test_app_default")
    @router.route!
    @router.layout.should == "default"
    @router.page.should == "about"
  end
  it "about layout yellow page" do
    @router = Router.new("/about/yellow","/index.php", File.dirname(__FILE__)+"/test_app_about")
    @router.route!
    @router.layout.should == "about"
    @router.page.should == "yellow"
  end
  it "about layout yellow page" do
    @router = Router.new("/animals/dogs/brian","/index.php", File.dirname(__FILE__)+"/test_app_about")
    @router.route!
    @router.layout.should == "default"
    @router.page.should == "dogs"
  end
end